src/Worker/Core/DurableTaskRegistry.cs

```diff
@@ -2,176 +2,73 @@
-            });
-        return this;
-    }
-
-    /// <summary>
-    /// Registers an orchestrator function that doesn't take any input nor returns any output.
-    /// </summary>
-    /// <inheritdoc cref="AddOrchestrator{TInput, TOutput}"/>
-    public DurableTaskRegistry AddOrchestrator(
-        TaskName name,
-        Func<TaskOrchestrationContext, Task> implementation)
-    {
-        Check.NotDefault(name);
-        Check.NotNull(implementation);
-        return this.AddOrchestrator<object?, object?>(name, async (ctx, _) =>
-        {
-            await implementation(ctx);
-            return null;
-        });
-    }
-
-    /// <summary>
-    /// Registers an orchestrator function that returns an output but doesn't take any input.
-    /// </summary>
-    /// <inheritdoc cref="AddOrchestrator{TInput, TOutput}"/>
-    public DurableTaskRegistry AddOrchestrator<TOutput>(
-        TaskName name,
-        Func<TaskOrchestrationContext, Task<TOutput>> implementation)
-    {
-        Check.NotDefault(name);
-        Check.NotNull(implementation);
-        return this.AddOrchestrator<object?, TOutput>(name, (ctx, _) => implementation(ctx));
-    }
-
-    /// <summary>
-    /// Registers an orchestrator function that takes an input and returns an output.
-    /// </summary>
-    /// <typeparam name="TInput">The orchestrator function's input type.</typeparam>
-    /// <typeparam name="TOutput">The orchestrator function's return type.</typeparam>
     /// <param name="name">The name of the orchestrator.</param>
-    /// <param name="implementation">The orchestration implementation as a function.</param>
-    /// <returns>Returns this <see cref="DurableTaskRegistry"/> instance.</returns>
-    public DurableTaskRegistry AddOrchestrator<TInput, TOutput>(
-        TaskName name,
-        Func<TaskOrchestrationContext, TInput, Task<TOutput>> implementation)
+    /// <param name="factory">The orchestrator factory.</param>
+    /// <returns>This registry instance, for call chaining.</returns>
+    /// <exception cref="ArgumentException">
+    /// Thrown if any of the following are true:
+    /// <list type="bullet">
+    ///   <item>If <paramref name="name"/> is <c>default</c>.</item>
+    ///   <item>If <paramref name="name" /> is already registered.</item>
+    ///   <item>If <paramref name="factory"/> is <c>null</c>.</item>
+    /// </list>
+    /// </exception>
+    public DurableTaskRegistry AddOrchestrator(TaskName name, Func<ITaskOrchestrator> factory)
     {
         Check.NotDefault(name);
-        Check.NotNull(implementation);
-        if (this.orchestratorsBuilder.ContainsKey(name))
+        Check.NotNull(factory);
+        if (this.activitiesBuilder.ContainsKey(name))
         {
-            throw new ArgumentException($"A task orchestrator named '{name}' is already added.", nameof(name));
+            throw new ArgumentException($"A TaskOrchestrator named '{name}' is already added.", nameof(name));
```

Comment:
If we're going to use actual class names, then we should probably use `nameof` instead of hardcoding them.

Reply:
Done